connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java

```diff
@@ -355,24 +357,44 @@ private WorkerTask buildWorkerTask(ConnectorConfig connConfig,
                                        Converter valueConverter) {
         // Decide which type of worker task we need based on the type of task.
         if (task instanceof SourceTask) {
+            // merging connector config with worker config to apply producer overrides
+            // on the connector
+            WorkerConfig mergedConfig = getMergedConfig(config, connConfig, "producer");
             TransformationChain<SourceRecord> transformationChain = new TransformationChain<>(connConfig.<SourceRecord>transformations());
             OffsetStorageReader offsetReader = new OffsetStorageReaderImpl(offsetBackingStore, id.connector(),
                     internalKeyConverter, internalValueConverter);
             OffsetStorageWriter offsetWriter = new OffsetStorageWriter(offsetBackingStore, id.connector(),
                     internalKeyConverter, internalValueConverter);
             KafkaProducer<byte[], byte[]> producer = new KafkaProducer<>(producerProps);
             return new WorkerSourceTask(id, (SourceTask) task, statusListener, initialState, keyConverter,
-                     valueConverter, transformationChain, producer, offsetReader, offsetWriter, config, time);
+                    valueConverter, transformationChain, producer, offsetReader, offsetWriter, mergedConfig, time);
         } else if (task instanceof SinkTask) {
+            // merging connector config with worker config to apply consumer overrides
+            // on the connector
+            WorkerConfig mergedConfig = getMergedConfig(config, connConfig, "consumer");
             TransformationChain<SinkRecord> transformationChain = new TransformationChain<>(connConfig.<SinkRecord>transformations());
-            return new WorkerSinkTask(id, (SinkTask) task, statusListener, initialState, config, keyConverter,
+            return new WorkerSinkTask(id, (SinkTask) task, statusListener, initialState, mergedConfig, keyConverter,
                     valueConverter, transformationChain, time);
         } else {
             log.error("Tasks must be a subclass of either SourceTask or SinkTask", task);
             throw new ConnectException("Tasks must be a subclass of either SourceTask or SinkTask");
         }
     }
 
+    // visible for testing
+    static WorkerConfig getMergedConfig(WorkerConfig workerConfig, ConnectorConfig connConfig, String prefix) {
+        Map<String, String> mergedConfigs = new HashMap<>(workerConfig.originals().size() + connConfig.originals().size());
+
+        mergedConfigs.putAll(workerConfig.originalsStrings());
+        mergedConfigs.putAll(connConfig.originalsStringsWithPrefixIntact(prefix));
```

Comment:
I'm a little uncertain about allowing overrides blindly. For example, I don't think we'd want to allow automatic offset commits to be enabled. On the other hand, it looks like we do apply worker-level configs blindly, so maybe it's fine.